### PR TITLE
Add textarea back and support IMEs

### LINKED ISSUE #124: Terminal doesn't accept input via IME
Using Ubuntu 16.04 native Korean IME the terminal just enters the English characters.

### PR DESCRIPTION
**Pull request by Tyriar**
Korean input using Korean (Hangul) (IBus) on Ubuntu 16.04

image

Japanese input using Japanese (Anthy) (IBus) on Ubuntu 16.04

image

image

Covered in this PR:

- [x] Manually revert: Drop hidden textarea for input handling #78
- [x] Support CSS focus state using the `.focus` class instead of a `:focus`
- [x] Support IMEs by hooking into composition events #124
- [x] Add a bunch of tests for `CompositionHelper`

Some of the testing that needs to be done on final version:

- [x] Copy using <kbd>ctrl</kbd>+<kbd>ins</kbd>
- [x] Paste using <kbd>shift</kbd>+<kbd>ins</kbd> and <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>v</kbd>
- [x] Inputting without IME continues to work as expected
- [x] Inputting via Korean IME
  - [x] Test: ㅇㅇ (automated)
  - [x] Test: 아아 (automated)
  - [x] Test: 앙앙 (automated)
  - [x] Pressing arrow keys to finalize composition
  - [x] Pressing enter to finalize and execute composition
  - [x] Numbers and punctuation can be input while IME is active (see remaining issues)
- [x] Inputting via Japanese IME
  - [x] Test multiple character compositions (automated)
  - [x] Test hiragana -> katakana conversion (automated)
  - [x] Test hiragana -> kanji conversion (automated)
- [x] Inputting non-composition character (eg. 1) after composition character (eg. ㅇ) (automated)
- [x] Check support on Windows (tested using Microsoft Korean IME and Google Japanese IME, works fine except for the fact that the terminal cursor disappears which makes the composition appear on top of text)
- [x] Check support on OS X (tested using Apple's 2-Set Korean IME)
- [x] Check browser support (seems to behave consistently across browsers)
  - [x] Chrome 51 (Linux, OS X, Windows)
  - [x] Firefox 48 (Linux)
  - [x] Safari 9.1.1 (OS X)

Fixes #124 (Terminal doesn't accept input via IME)

Remaining issues:

- Currently the composition view is placed on top of the cursor, it's not aligned currently due to #149 (Lines containing unicode characters differ in height)
- The composition view can appear on top of text on Windows only because the cursor seems to disappear. Input works fine, it's just that it can overlap on text.

### CODE REVIEW

**Comment by Tyriar** [comment]
@parisk please review when you get a chance, so far I've only tested this on Korean but it's probably the most complex. I'll be able to test multi-character compositions with Japanese but it will likely be small tweaks inside `CompositionHelper` to get this working if it's not already.

The code itself is very well sectioned off from the rest so you don't have to worry so much about it introducing instability when not using an IME. A section is added to the top of `Terminal.prototype.keyDown` which prevents the key event going through if the `isComposing` flag is active. There is also a call at the end of refresh to position the composition view element on top of the cursor. All other changes not in `CompositionHelper` are just reverting the textarea change.

**Comment by parisk** [comment]
Thanks a lot for this @Tyriar! Right now I will focus on what is being needed to release **1.0.0** by the end of the week (e.g. adding a CONTRIBUTING.md file, updating the README, etc.) and I will get on top of this starting next week at most.

**Comment by Tyriar** [comment]
@parisk thanks, probably a good idea to release v1 before pushing a change this big. If we can land this within a week it will get a bunch of testing in our Insider's build before the stable release.

**Comment by Tyriar** [comment]
Ok done and tested, at least under Chrome 51 and Ubuntu 16.04. I moved even more into the `CompositionHelper` so now the only changes made outside are:

- Creating the `CompositionHelper`
- Attaching events to it
- Calling into `CompositionHelper.prototype.keydown` manually at the beginning of `Terminal.prototype.keyDown`

**Comment by yuvipanda** [comment]
(I can also help test this at paws.wmflabs.org once it lands!)

**Comment by Tyriar** [comment]
@parisk good to go :+1: the only remaining issue is the IME helper can popup in the wrong spot on Windows due to the cursor disappearing. The fix for that likely lies in pty.js or winpty though.

**Comment by parisk** [comment]
Thanks a lot @Tyriar! Moving on with testing on OS X.

File: src/xterm.js [hunk 1]
```diff
@@ -757,11 +867,39 @@
...
+      this.helperContainer.classList.add('xterm-helpers');
+      // TODO: This should probably be inserted once it's filled to prevent an additional layout
+      this.element.appendChild(this.helperContainer);
+      this.textarea = document.createElement('textarea');
+      this.textarea.classList.add('xterm-helper-textarea');
+      this.textarea.setAttribute('autocorrect', 'off');
+      this.textarea.setAttribute('autocapitalize', 'off');
+      this.textarea.setAttribute('spellcheck', 'false');
+      this.textarea.tabIndex = 0;
+      this.textarea.onfocus = function() {
```
Comment (parisk) [inline src/xterm.js]: Could you please use `this.textarea.addEventListener('focus, ...)` instead here please?

File: src/xterm.js [hunk 2]
```diff
@@ -757,11 +867,39 @@
...
+        self.emit('focus', {terminal: self});
+      }
+      this.textarea.onblur = function() {
```
Comment (parisk) [inline src/xterm.js]: Could you please use `this.textarea.addEventListener('blur', ...)` instead here please?

File: src/xterm.js [hunk 3]
```diff
@@ -436,7 +625,18 @@
      * Focus the terminal. Delegates focus handling to the terminal's DOM element.
      */
     Terminal.prototype.focus = function() {
-      return this.element.focus();
+      if (document.activeElement === this.textarea) {
+        return;
+      }
+
+      if (this.sendFocus) {
+        this.send('\x1b[I');
```
Comment (parisk) [inline src/xterm.js]: This is already done inside the `focus` callback that is set up in the `Terminal.bindFocus` static method. I would prefer to keep any "business logic" inside high level callbacks. Defaults can be set up in `bindFocus` which is run on every new instance.

File: src/xterm.js [hunk 4]
```diff
@@ -460,7 +660,18 @@
...
+      if (Terminal.focus !== this) {
+        return;
+      }
+
+      this.element.classList.remove('focus');
+      this.cursorState = 0;
+      this.refresh(this.y, this.y);
+      this.textarea.blur();
+      if (this.sendFocus) {
+        this.send('\x1b[0]');
```
Comment (parisk) [inline src/xterm.js]: This is already done inside the callback that is set up in the `Terminal.bindBlur` static method. I would prefer to keep any "business logic" inside high level callbacks. Defaults can be set up in `bindBlur` which is run on every new instance.

**Comment by parisk** [comment]
🆗 , I made my first comments on code styling and I have a few more on functionality on OS X.

1. **Cannot paste using right click**. Quick hack proposal: Reposition the helper textarea below the mouse position on `mousedown` if the right click is being pressed. This should trigger the `contextmenu` event on the textarea and the user should be able to paste using right click.
2. **Cannot enter emojis**. Xterm ignores emojis completely when attempting to enter them with the native input method on OS X. Well this is **definitely non-blocking** for merging this PR (no matter how much I love emojis in the terminal 😍 ), but thought about mentioning.

 In general this PR behaves really well and is exceptionally written and documented! 👍  @Tyriar!

File: src/xterm.js (hunk 4 above)
Comment (Tyriar) [inline src/xterm.js]: See https://github.com/sourcelair/xterm.js/pull/175/commits/4595a18197a9f7378077b078c06719ecb16adb38, let me know if that's what you meant.

**Comment by Tyriar** [comment]
@parisk so I took a shot at fixing the paste on right click behavior and while I got it to work by positioning the textarea under the cursor and then redirecting the event to the textarea and cancelling the original event, it removes the selection which breaks means you cannot copy using right click. Here's the super messy/hacky code I used to accomplish this:

```
--- a/src/xterm.js
+++ b/src/xterm.js
@@ -1220,7 +1220,41 @@
         };
       }
 
+function getOffset( el ) {
+    var _x = 0;
+    var _y = 0;
+    while( el && !isNaN( el.offsetLeft ) && !isNaN( el.offsetTop ) ) {
+        _x += el.offsetLeft - el.scrollLeft;
+        _y += el.offsetTop - el.scrollTop;
+        el = el.offsetParent;
+    }
+    return { top: _y, left: _x };
+}
+
+self.textarea.addEventListener('mousedown', function(ev) {
+  console.log('textarea mousedown', ev);
+ev.stopPropagation();
+});
+
       on(el, 'mousedown', function(ev) {
+
+        self.textarea.style.position = 'absolute';
+var offset = getOffset(self.element);
+        self.textarea.style.left = (ev.pageX - offset.left - 10) + 'px';
+        self.textarea.style.top = (ev.pageY - offset.top - 10) + 'px';
+self.textarea.style.width = '20px';
+self.textarea.style.height = '20px';
+self.textarea.style.zIndex = '0';
+        redirectMouseEvent(ev, self.textarea);
+
+function redirectMouseEvent(event, el) {
+  el.dispatchEvent(new MouseEvent(event.type, event));
+}
+ev.preventDefault();
+    ev.stopPropagation();
+    return;
+
+
         if (!self.mouseEvents) return;
 
         // send the button
```

Perhaps this should be deferred until selection is handled in a non-native way which would fix #68 and #69. I think that's how hterm behaves and it's how the VS Code editor maintains selection no matter where the focus is, seems like the right solution without compromising.

There is always <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>v</kbd> and <kbd>cmd</kbd>+<kbd>v</kbd> to paste now which are the defaults on Linux/Mac.

I'd prefer to get this in and accept the regression with the context menu for now as this change needs to get in for VS Code 1.4 (code freeze/testing starts on the 25th).

**Comment by parisk** [comment]
🆗  let's move forward with merging this right now and I will create an issue in order to fix copy and paste with mouse until the next release (1.1.0) at 22 August 2016.

**Comment by Tyriar** [comment]
https://github.com/sourcelair/xterm.js/pull/175/commits/4595a18197a9f7378077b078c06719ecb16adb38 actually broke the IMEs as it changed focus/blur behavior on the textarea, investigating...

**Pull request merged**